fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/EntityElytraEvents.java

```diff
@@ -0,0 +1,116 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.entity.event.v1;
+
+import net.minecraft.entity.EquipmentSlot;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.item.ElytraItem;
+import net.minecraft.item.ItemStack;
+import net.minecraft.world.event.GameEvent;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+/**
+ * Events related to elytra flight for living entities. Elytra flight is also knows as "fall flying".
+ */
+public class EntityElytraEvents {
+	/**
+	 * An event that allows listeners to grant elytra flight to living entities.
+	 * Will be called when players try to start elytra flight by pressing space in mid-air, and every tick for all living entities to check if elytra flight is still allowed.
+	 *
+	 * <p>Items that wish to grant elytra flight when worn in the chest equipment slot can simply implement {@link FabricElytraItem} instead of registering a listener.
+	 */
+	public static final Event<Grant> GRANT = EventFactory.createArrayBacked(Grant.class, listeners -> (livingEntity, damageElytra) -> {
+		for (Grant listener : listeners) {
+			if (listener.grantElytraFlight(livingEntity, damageElytra)) {
+				return true;
+			}
+		}
+
+		return false;
+	});
+
+	static {
+		GRANT.register((livingEntity, damageElytra) -> {
+			ItemStack chestStack = livingEntity.getEquippedStack(EquipmentSlot.CHEST);
+
+			if (chestStack.getItem() instanceof FabricElytraItem fabricElytraItem) {
+				return fabricElytraItem.grantElytraFlight(livingEntity, chestStack, damageElytra);
+			}
+
+			return false;
+		});
+	}
+
+	@FunctionalInterface
+	public interface Grant {
+		/**
+		 * Check if elytra flight should be granted to an entity.
+		 *
+		 * @param livingEntity the entity
+		 * @param tickElytra   false if this is just to check if elytra flight is allowed, true if the elytra should be ticked, i.e. damaged.
+		 * @return true to grant elytra flight, and cancel subsequent handlers
+		 */
+		boolean grantElytraFlight(LivingEntity livingEntity, boolean tickElytra);
+	}
+
+	/**
+	 * An interface that can be implemented on an item to enable elytra flight when it is worn in the {@link EquipmentSlot#CHEST} slot.
+	 *
+	 * <p>To disable cape rendering when this item is worn (like the vanilla elytra item), have a look at {@code LivingEntityFeatureRenderEvents}.
+	 */
+	public interface FabricElytraItem {
```

Comment:
> I'd recommend this not be an inner class, but that's purely for aesthetic reasons of preferring `implements FabricElytraItem` over `implements EntityElytraEvents.FabricElytraItem`.

This was done to avoid spamming API classes for something that is essentially a convenient implementation of the event.

> Also, since this itself isn't really an `Item`, shouldn't this just be called `FabricElytra` and not `FabricElytraItem`?

It's not an item, but it only works if it's implemented on an `Item` subclass, so implementors should be items. Another advantage is that people searching for `ElytraItem` should find this class in the popup.

Does this explain the decisions?

Reply:
I agree with Shnupbups here (esp. about the nesting). `implements EntityElytraEvents.FabricElytraItem` is honestly pretty ugly and having a top-level interface would be nicer.